docs/control-structures.rst

```diff
@@ -488,3 +491,71 @@ In the above example, ``revert("Not enough Ether provided.");`` returns the foll
 .. note::
     There used to be a keyword called ``throw`` with the same semantics as ``revert()`` which
     was deprecated in version 0.4.13 and removed in version 0.5.0.
+
+``try``/``catch``
+-----------------
+
+A failure in an external call can be caught using a try/catch statement, as follows:
+
+::
+
+    pragma solidity ^0.6.0;
+
+    interface DataFeed { function getData() public returns (uint value); }
+
+    contract FeedConsumer {
+        DataFeed feed;
+        uint errorCount;
+        function rate(address token) public returns (uint value, bool success) {
+            // Permanently disable the mechanism if there are
+            // more than 10 errors.
+            require(errorCount < 10);
+            try feed.getData() returns (uint value) {
+                return (value, true);
+            } catch Error(string memory reason) {
+                // This is executed in case
+                // revert was called inside getData
+                // and a reason string was provided.
+                errorCount++;
+                return (0, false);
+            } catch (bytes memory lowLevelData) {
+                // This is executed in case there
+                // was a failing assertion, division
+                // by zero, etc.
+                errorCount++;
+                return (0, false);
+            }
+        }
+    }
+
+After the ``try`` keyword, there has to be an expression that evaluates to an external call.
+Any error inside the expression itself is not caught, only the error returned by the
+external call itself. The ``returns`` part that follows declares return variables
+matching the types returned by the external call. In case there was no error,
+these variables are assigned and the contract's execution continues inside the
+first success block. If the end of the success block is reached, execution continues
+after the ``catch`` blocks.
+
+Currently, Solidity supports different kinds of catch blocks depending on the
+type of error. If the error was caused by ``revert("reasonString")`` or
+``require(false, "reasonString")``, then the catch clause
+of the type ``catch Error(string memory reason)`` will be executed.
+
+It is planned to support other types of error data in the future.
+The string ``Error`` is currently parsed as is and is not treated as an identifier.
+
+The clause ``catch (bytes memory lowLevelData)`` is executed if the error signature
+does not match any other clause or if no error data was provided with the exception.
+The declared variable provides access to the low-level error data in that case.
+
+If you are not interested in the error data, you can just use
+``catch ()`` (even as the only catch clause).
```

Comment:
Should this be ``catch () { ... }``  or ``catch { ... }`` :-)?

Reply:
Since we also have `function f() { ... }`, I would say `catch () { ... }`.

I'm wondering whether we should not require `try (x.e()) returns ( ...)` since we also have `if (x.e()) { ... }`.

Reply:
Also, is the `returns ( ... )` part optional?

Reply:
I was wondering because ``catch()`` might seem like only catching empty error data not all kinds of error data... We could also make it ``catch(...)`` instead of ``catch``, but maybe ``catch()`` is just fine, I'm not sure.

The ``if`` condition is different in that it can be complex expressions, where here we just have a single call, right? It might be ``try (return_me_a_function()).x()`` or ``try (cond ? f1 : f2)()``, etc., but the whole thing has to be a call in the end, right? That's why I'd actually not even not require, but maybe even disallow outer parentheses there - I think that might fit better with the ``returns`` after it, but I'm not entirely sure.

Hm... if ``returns(...)`` is optional, would there be a different semantics for ``try x.e() returns () { ... }`` and ``try x.e() { ... }``? In that the first requires empty returndata and the second just ignores returndata? Phew this is indeed a whole lot of stuff to consider :-).

Reply:
catch: Would you disallow `catch ()`? Note that this is different from `catch Error()`.

Agree on the parentheses. Yes, the expression type has to be an external call, nothing else.

Missing returns is the same as `f()` instead of `x = f()`. I think it still requires properly abi-formatted return data, regardless of whether it is used or not.

Reply:
I'm not yet sure about missing returns, but maybe you're right.

catch: It might be best to disallow ``catch ()``, because it could either be "catch all" or "catch empty"... I'd suggest to keep it at ``catch Error(...) {...}``, ``catch(bytes memory) {...}`` and ``catch {...}`` (similarly to ``else`` in ``if(...) ... else if(...) ... else ...``) - I think then everything is well defined and clear and whoever wants can distinguish empty from non-empty in using ``msg.length`` in ``catch(bytes memory msg)``.